syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package gfs;

// The GFS service definition.
service GFS {
  // Client -> Chunkserver RPCs
  // --------------------------

  // Read data
  rpc ReadChunk (ReadChunkRequest) returns (ReadChunkReply) {}
  // Write data
  rpc WriteChunk (WriteChunkRequest) returns (WriteChunkReply) {}
  // Sends a ping
  rpc ClientServerPing (PingRequest) returns (PingReply) {}

  // TODO: AppendChunk
  // TODO: PushData

  // Master -> Chunkserver RPCs
  // --------------------------

  // Delete chunks when master knows it is no longer needed.
  rpc DeleteChunks (DeleteChunksRequest) returns (DeleteChunksReply) {}

  // Chunkserver -> Chunkserver RPCs
  // --------------------------

  // TODO: SerializedWrite (or similar)
}

// The request message containing the user's name.
message PingRequest {
  string name = 1;
}

// The response message containing the greeting
message PingReply {
  string message = 1;
}

// Read Chunk Request
message ReadChunkRequest {
  int64 chunkhandle = 1;
  int64 offset = 2;
  int64 length = 3;
}

// Write Chunk Request
message WriteChunkRequest {
  int64 chunkhandle = 1;
  string data = 2;
  int64 offset = 3;
// TODO: hash of the data
}

// Read Chunk Reply
message ReadChunkReply {
  string data = 1;
  int64 bytes_read = 2;
// TODO: hash of the data
}

// Write Chunk Reply
message WriteChunkReply {
  int64 bytes_written = 1;
}

message DeleteChunksRequest {
  // List of chunkhandles to delete.
  repeated int64 chunkhandles = 1;
}

message DeleteChunksReply {}

service GFSMaster {
  // Client -> Master RPCs
  // ---------------------
    
  // List all files with filename matching a certain prefix.
  // - Directories are part of the filename, e.g. "abc/def/ghi.txt"
  rpc FindMatchingFiles (FindMatchingFilesRequest) returns (FindMatchingFilesReply) {}

  // For a given filename and chunk index, returns the chunk handle and locations of replicas.
  // - Does not attempt to create the chunk.
  // - Should be used for read requests.
  rpc FindLocations (FindLocationsRequest) returns (FindLocationsReply) {}

  // For a given filename and chunk index, returns the chunk handle and locations of replicas.
  // - Will create the chunk if it does not exists (this replaces AddChunk).
  // - Also identifies the primary (lease holder).
  // - (in actual GFS) For chunks that are copy-on-write, this will perform the copy.
  // - Should be used for write requests.
  rpc FindLeaseHolder (FindLeaseHolderRequest) returns (FindLeaseHolderReply) {}

  // Requests master to move (rename) the file.
  rpc MoveFile (MoveFileRequest) returns (MoveFileReply) {}

  // Requests master to delete file.
  rpc DeleteFile (DeleteFileRequest) returns (DeleteFileReply) {}

  // For a given filename and chunk index, allocate chunk servers, returns the chunk handle and
  // locations of replicas.
  rpc AddChunk (AddChunkRequest) return (AddChunkReply) {}

  // For a given filename, returns the length of the file.
  rpc GetFileLength (GetFileLengthRequest) return (GetFileLengthReply) {}

  // Chunkserver -> Master RPCs
  // --------------------------

  // Chunkserver should periodically send heartbeats to the master, once every few seconds.
  // - If the master does not receive a heartbeat for some time, it will assume failure.
  // - When master restarts, it will send this to get the chunks.
  rpc Heartbeat (HeartbeatRequest) returns (HeartbeatReply) {}

  // Master -> Master RPCs
  // ---------------------

  // TODO: RPC to replicate data to shadow masters.
}

message FindMatchingFilesRequest {
  string prefix = 1;
}

// Status returned will be OK (or a network error).
message FindMatchingFilesReply {
  message File {
    string filename = 1;
    // The chunk indexes range from 0 to num_chunks-1.
    int64 num_chunks = 2;
  }
  repeated File files = 1;
}

message FindLocationsRequest {
  string filename = 1;
  int64 chunk_index = 2;
}

message FindLocationsReply {
  int64 chunkhandle = 1;
  // Each location is a string "IP:port" like "127.0.0.1:12345".
  repeated string locations = 2;
}

message FindLeaseHolderRequest {
  string filename = 1;
  int64 chunk_index = 2;
}

// Status returned may be OK or NOT_FOUND (or a network error).
message FindLeaseHolderReply {
  int64 chunkhandle = 1;
  // Each location is a string "IP:port" like "127.0.0.1:12345".
  repeated string locations = 2;
  // The primary location is also "IP:port". It will be also present in locations.
  string primary_location = 3;
}

message MoveFileRequest {
  string old_filename = 1;
  string new_filename = 2;
}

// Status returned may be OK or NOT_FOUND (or a network error).
message MoveFileReply {}

message DeleteFileRequest {
  string filename = 1;
}

// Status returned may be OK or NOT_FOUND (or a network error).
message DeleteFileReply {}

message AddChunkRequest {
  string filename = 1;
  int64 chunk_index = 2;
}

message AddChunkReply {
  // TODO: define the reply
}

message GetFileLengthRequest {
  string filename = 1;
}

message GetFileLengthReply {
  int64 filen_length = 1;
}

message HeartbeatRequest {
  message ChunkInfo {
    int64 chunkhandle = 1;
    // TODO int64 version = 2;
  }
  // All chunks stored in chunkserver.
  repeated ChunkInfo chunks = 1;
  // The location is a string "IP:port" like "127.0.0.1:12345".
  // - It should be the location of the chunkserver's GRPC server.
  string location = 2;
}

// Status returned will be OK (or a network error).
message HeartbeatReply {}
